be/src/exec/pipeline/pipeline_driver_queue.h

```diff
@@ -49,7 +70,50 @@ class SubQuerySharedDriverQueue {
 
     double accu_time_after_divisor() { return _accu_consume_time.load() / factor_for_normal; }
 
-    std::queue<DriverRawPtr> queue;
+    void put(const DriverRawPtr driver) {
+        if (driver->driver_state() == DriverState::CANCELED) {
+            queue.emplace_front(driver);
```

Comment:
Why not directly put the driver into cancelled_set?

Reply:
> Why not directly put the driver into cancelled_set?

let me explain the reasons for this design in detail. 

After finalize, the lifecycle of driver ends with the fragment context. The DriverQueue holds the DriverRawPtrs, which may be invalid when they are taken from DriverQueue. So I introduced `cancelled_set` for verification. A Driver in the cancelled_set, which means that it has been taken from `pending_cancel_queue` but remains in `queue`. A premise of this design is that the driver in `pending_cancel_queue` must be in `queue`.

You may also want to ask why not directly put the driver into `pending_cancel_queue`. We add driver into `cancelled_set` when taking it from `pending_cancel_queue` and delete if from `cancelled_set` when taking it from `queue`. If a driver that is not in the `queue` appears in `pending_cancel_queue`, it will never be delete from `cancelled_set`.